File: app/src/main/java/org/michaelbel/moviemade/presentation/features/similar/SimilarMoviesPresenter.java
package org.michaelbel.moviemade.presentation.features.similar;

import org.jetbrains.annotations.NotNull;
import org.michaelbel.moviemade.core.utils.Error;
import org.michaelbel.moviemade.core.utils.NetworkUtil;

import io.reactivex.disposables.CompositeDisposable;
import retrofit2.HttpException;

public class SimilarMoviesPresenter implements SimilarContract.Presenter {

    private int page;
    private SimilarContract.View view;
    private SimilarContract.Repository repository;
    private CompositeDisposable disposables = new CompositeDisposable();

    public SimilarMoviesPresenter(SimilarRepository repository) {
        this.repository = repository;
    }

    @Override
    public void attach(@NotNull SimilarContract.View view) {
        this.view = view;
    }

    @Override
    public void getSimilarMovies(int movieId) {
        if (!NetworkUtil.INSTANCE.isNetworkConnected()) {
            view.setError(new Throwable("No connection"), Error.ERR_NO_CONNECTION);
            return;
        }

        page = 1;
        disposables.add(repository.getSimilarMovies(movieId, page)
            .subscribe(
                response -> {
                    if (response.getMovies().isEmpty()) {
                        view.setError(new Throwable("No movies"), Error.ERR_NO_MOVIES);
                        return;
                    }
                    view.setMovies(response.getMovies());
                },
                e -> view.setError(e, ((HttpException) e).code())
            )
        );
    }

    @Override
    public void getSimilarMoviesNext(int movieId) {
        if (!NetworkUtil.INSTANCE.isNetworkConnected()) {
            return;
        }

        page++;
        disposables.add(repository.getSimilarMovies(movieId, page)
            .subscribe(response -> view.setMovies(response.getMovies())));
    }

    @Override
    public void destroy() {
        disposables.dispose();
    }
}

File: app/src/main/java/org/michaelbel/moviemade/presentation/features/similar/SimilarMoviesPresenter.kt
package org.michaelbel.moviemade.presentation.features.similar

import org.michaelbel.moviemade.core.utils.Error
import org.michaelbel.moviemade.core.utils.NetworkUtil
import org.michaelbel.moviemade.presentation.base.Presenter
import retrofit2.HttpException

class SimilarMoviesPresenter(repository: SimilarRepository): Presenter(), SimilarContract.Presenter {

    private var page: Int = 0
    private var view: SimilarContract.View? = null
    private val repository: SimilarContract.Repository = repository

    override fun attach(view: SimilarContract.View) {
        this.view = view
    }

    override fun getSimilarMovies(movieId: Int) {
        if (!NetworkUtil.isNetworkConnected()) {
            view?.setError(Throwable("No connection"), Error.ERR_NO_CONNECTION)
            return
        }

        page = 1
        disposable.add(repository.getSimilarMovies(movieId, page)
                .subscribe({
                    if (it.movies.isEmpty()) {
                        view?.setError(Throwable("No movies"), Error.ERR_NO_MOVIES)
                        return@subscribe
                    }
                    view?.setMovies(it.movies)
                }, { view?.setError(it, (it as HttpException).code()) })
        )
    }

    override fun getSimilarMoviesNext(movieId: Int) {
        if (!NetworkUtil.isNetworkConnected()) {
            return
        }

        page++
        disposable.add(repository.getSimilarMovies(movieId, page)
                .subscribe { view?.setMovies(it.movies) })
    }

    override fun destroy() {
        disposable.dispose()
    }
}